# Building a Spam Filter with Naive Bayes

In this project we are aiming to build a spam filter for SMS messages using the multinomial Naive Bayes algorithm and a dataset of 5,572 SMS messages aleready classified by humans. The dataset ([link](https://archive.ics.uci.edu/ml/datasets/sms+spam+collection)) was put together by Tiago A. Almeida and José María Gómez Hidalgo, and the data collection process is decribed on [this page](http://www.dt.fee.unicamp.br/~tiago/smsspamcollection/#composition).

## Exploring the Dataset

To start, we will read in the dataset and do some exploration.

In [13]:
import pandas as pd
spam_collection = pd.read_csv('SMSSpamCollection', sep = '\t', header = None, names = ['Label', 'SMS'])
spam_collection.size
spam_collection.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
Label    5572 non-null object
SMS      5572 non-null object
dtypes: object(2)
memory usage: 87.1+ KB


In [14]:
#finding what percentage  of messages is spam and not-spam
spam_collection['Label'].value_counts()

ham     4825
spam     747
Name: Label, dtype: int64

We found out that spam messages take up to 13% of the total messages, and 87% of messages are ham ('ham' means non-spam).


## Training and Test Set


Before creating the spam filter we would need to design the test first in order to make sure  the software (the spam filter) is working correctly.

For this purpose, we will divide our dataset into two categories: 

* A training set;
* A test set.

We will use 80% of the dataset for training and remaining 20% for testing.
As we know, that all messages in our dataset are classifies by human, on a test stage we will treat those 20% of messages as new and have the filter classify them. After test is done, we will be able to compare the algorithm classification with the human classificaiton.

Our goal will be to create a spam filter that classifies new messages with an accuracy greater that 80%.

In [15]:
#Randomizing the entore dataset so spam and ham are spread properly
data_randomized = spam_collection.sample(frac = 1, random_state= 1)
data_randomized

Label                                                SMS
1078   ham                       Yep, by the pretty sculpture
4028   ham      Yes, princess. Are you going to make me moan?
958    ham                         Welp apparently he retired
4642   ham                                            Havent.
4674   ham  I forgot 2 ask ü all smth.. There's a card on ...
5461   ham  Ok i thk i got it. Then u wan me 2 come now or...
4210   ham  I want kfc its Tuesday. Only buy 2 meals ONLY ...
4216   ham                         No dear i was sleeping :-P
1603   ham                          Ok pa. Nothing problem:-)
1504   ham                    Ill be there on  &lt;#&gt;  ok.
1783   ham  My uncles in Atlanta. Wish you guys a great se...
3465   ham                                           My phone
5534   ham                       Ok which your another number
4267   ham  The greatest test of courage on earth is to be...
2498   ham  Dai what this da.. Can i send my resume to thi...
4259   ham                      I am late. I will be there at
147   spam  FreeMsg Why haven't you replied to my text? I'...
141    ham                  K, text me when you're on the way
4517  spam  Congrats! 2 mobile 3G Videophones R yours. cal...
3053   ham  Please leave this topic..sorry for telling that..
5392   ham  Ooooooh I forgot to tell u I can get on yovill...
2346   ham          Hi this is yijue, can i meet u at 11 tmr?
1242   ham  I want to show you the world, princess :) how ...
3224   ham                  Well that must be a pain to catch
4872   ham                Well. You know what i mean. Texting
3044   ham         Your bill at 3 is £33.65 so thats not bad!
1660   ham                       Yeah, where's your class at?
3214   ham                                     What's ur pin?
501    ham  Fighting with the world is easy, u either win ...
1827   ham  Dude. What's up. How Teresa. Hope you have bee...
...    ...                                                ...
1031   ham  Can not use foreign stamps in this country. Go...
1110   ham                    S s..first time..dhoni rocks...
1888  spam  Urgent! Please call 09061743811 from landline....
3550   ham  I got like $ &lt;#&gt; , I can get some more l...
1527   ham  Wow ... I love you sooo much, you know ? I can...
753    ham                        Dont gimme that lip caveboy
3049   ham           Die... Now i have e toot fringe again...
2628   ham  I know I'm lacking on most of this particular ...
562    ham                   Thanx 4 e brownie it's v nice...
4764   ham                         Prepare to be pleasured :)
3562  spam  Text BANNEDUK to 89555 to see! cost 150p texto...
252    ham  Wen ur lovable bcums angry wid u, dnt take it ...
2516   ham  Bognor it is! Should be splendid at this time ...
2962   ham  I'm doing da intro covers energy trends n pros...
4453   ham  I've told you everything will stop. Just dont ...
5374   ham  Do u konw waht is rael FRIENDSHIP Im gving yuo...
5396   ham           As in i want custom officer discount oh.
1202   ham                               I know she called me
3462   ham  K.. I yan jiu liao... Sat we can go 4 bugis vi...
2797   ham  Tell your friends what you plan to do on Valen...
4225   ham  Double eviction this week - Spiral and Michael...
144    ham         I know you are. Can you pls open the back?
5056   ham  Am on a train back from northampton so i'm afr...
2895   ham                   K...k...yesterday i was in cbe .
2763   ham  ARR birthday today:) i wish him to get more os...
905    ham  We're all getting worried over here, derek and...
5192   ham  Oh oh... Den muz change plan liao... Go back h...
3980   ham  CERI U REBEL! SWEET DREAMZ ME LITTLE BUDDY!! C...
235   spam  Text & meet someone sexy today. U can find a d...
5157   ham                            K k:) sms chat with me.

[5572 rows x 2 columns]

In [16]:
#Splitting the randomized dataset into training and test set
training_test_index = round(len(data_randomized) * 0.8)

# Training/Test split
training_set = data_randomized[:training_test_index].reset_index(drop=True)
test_set = data_randomized[training_test_index:].reset_index(drop=True)

print(training_set.shape)
print(test_set.shape)


print('training_set:', '\n',training_set['Label'].value_counts(),'\n')
print('test_set:','\n', test_set['Label'].value_counts())

(4458, 2)
(1114, 2)
training_set: 
 ham     3858
spam     600
Name: Label, dtype: int64 

test_set: 
 ham     967
spam    147
Name: Label, dtype: int64


After we randomized the entire dataset and split it, we calculated that spam percentage is 13% for both training and test sets.

## Letter Case and Punctuation


Before we use the training set to teach the algorithm to classify new messages, we will have to perform a bit of data cleaning to bring the data in a format that will allow us to extract easily all the information we need. We will have make all words in lower case and get rid of any punctuation.

In [17]:
# Before cleaning
training_set.head()

Label                                                SMS
0   ham                       Yep, by the pretty sculpture
1   ham      Yes, princess. Are you going to make me moan?
2   ham                         Welp apparently he retired
3   ham                                            Havent.
4   ham  I forgot 2 ask ü all smth.. There's a card on ...

In [18]:
#Removing punctuaiton
training_set['SMS'] = training_set['SMS'].str.replace('\W', ' ')
#Transforming to lower case
training_set['SMS'] = training_set['SMS'].str.lower()
training_set.head()

Label                                                SMS
0   ham                       yep  by the pretty sculpture
1   ham      yes  princess  are you going to make me moan 
2   ham                         welp apparently he retired
3   ham                                            havent 
4   ham  i forgot 2 ask ü all smth   there s a card on ...

## Creating the Vocabulary

Let's create a list with all of the unique words that occier in the messages of the training set.

In [20]:
#Transforming each message into a list 
training_set['SMS'] = training_set['SMS'].str.split()

vocabulary = []
for sms in training_set['SMS']:
    for word in sms:
        vocabulary.append(word)
        
vocabulary = list(set(vocabulary))

It looks like there are 7,783 unique words in all the messages of our training set.

In [21]:
len(vocabulary)

7783

## The Final Training Set

At this stage we will transform each unque word og the vocabulary to separate columns of our tarining set and how how many times each word was used.

In [22]:
word_counts_per_sms = {unique_word: [0] * len(training_set['SMS']) for unique_word in vocabulary}

for index, sms in enumerate(training_set['SMS']):
    for word in sms:
        word_counts_per_sms[word][index] += 1
        
word_counts = pd.DataFrame(word_counts_per_sms)
word_counts.head()        

0  00  000  000pes  008704050406  0089  01223585334  02  0207  02072069400  \
0  0   0    0       0             0     0            0   0     0            0   
1  0   0    0       0             0     0            0   0     0            0   
2  0   0    0       0             0     0            0   0     0            0   
3  0   0    0       0             0     0            0   0     0            0   
4  0   0    0       0             0     0            0   0     0            0   

  ...  zindgi  zoe  zogtorius  zouk  zyada  é  ú1  ü  〨ud  鈥  
0 ...       0    0          0     0      0  0   0  0    0  0  
1 ...       0    0          0     0      0  0   0  0    0  0  
2 ...       0    0          0     0      0  0   0  0    0  0  
3 ...       0    0          0     0      0  0   0  0    0  0  
4 ...       0    0          0     0      0  0   0  2    0  0  

[5 rows x 7783 columns]

In [23]:
training_set_clean = pd.concat([training_set, word_counts], axis=1)
training_set_clean.head()

Label                                                SMS  0  00  000  \
0   ham                  [yep, by, the, pretty, sculpture]  0   0    0   
1   ham  [yes, princess, are, you, going, to, make, me,...  0   0    0   
2   ham                    [welp, apparently, he, retired]  0   0    0   
3   ham                                           [havent]  0   0    0   
4   ham  [i, forgot, 2, ask, ü, all, smth, there, s, a,...  0   0    0   

   000pes  008704050406  0089  01223585334  02 ...  zindgi  zoe  zogtorius  \
0       0             0     0            0   0 ...       0    0          0   
1       0             0     0            0   0 ...       0    0          0   
2       0             0     0            0   0 ...       0    0          0   
3       0             0     0            0   0 ...       0    0          0   
4       0             0     0            0   0 ...       0    0          0   

   zouk  zyada  é  ú1  ü  〨ud  鈥  
0     0      0  0   0  0    0  0  
1     0      0  0   0  0    0  0  
2     0      0  0   0  0    0  0  
3     0      0  0   0  0    0  0  
4     0      0  0   0  2    0  0  

[5 rows x 7785 columns]

## Calculating Constants First

Now that we are done with data cleaning, we can begin creating the spam filter. The Naive Bayes algorithm will need to answer these two probability questions to be able to classify new messages:

$$
P(Spam | w_1,w_2, ..., w_n) \propto P(Spam) \cdot \prod_{i=1}^{n}P(w_i|Spam)
$$

$$P(Ham | w_1,w_2, ..., w_n) \propto P(Ham) \cdot\prod_{i=1}^{n}P(w_i|Ham)$$

Also, to calculate P(wi|Spam) and P(wi|Ham) inside the formulas above, we'll need to use these equations:

$$
P(w_i|Spam) = \frac{N_{w_i|Spam} + \alpha}{N_{Spam} + \alpha \cdot N_{Vocabulary}}
$$

$$
P(w_i|Ham) = \frac{N_{w_i|Ham} + \alpha}{N_{Ham} + \alpha \cdot N_{Vocabulary}}
$$

Some of the terms in the four equations above will have the same value for every new message. We can calculate the value of these terms once and avoid doing the computations again when a new messages comes in. Below, we'll use our training set to calculate:

* P(Spam) and P(Ham)
* NSpam, NHam, NVocabulary

We'll also use Laplace smoothing and set $\alpha = 1$.


In [24]:
# Isolating spam and ham messages first
spam_messages = training_set_clean[training_set_clean['Label'] == 'spam']
ham_messages = training_set_clean[training_set_clean['Label'] == 'ham']

# P(Spam) and P(Ham)
p_spam = len(spam_messages) / len(training_set_clean)
p_ham = len(ham_messages) / len(training_set_clean)

# N_Spam
n_words_per_spam_message = spam_messages['SMS'].apply(len)
n_spam = n_words_per_spam_message.sum()

# N_Ham
n_words_per_ham_message = ham_messages['SMS'].apply(len)
n_ham = n_words_per_ham_message.sum()

# N_Vocabulary
n_vocabulary = len(vocabulary)

# Laplace smoothing
alpha = 1

## Calculating Parameters

After we have calculated the constants that will have constant values in our equations for every new message, we can start working on $
P(w_i|Spam)$ and $P(w_i|Ham)$. Each parameter will thus be a conditional probability value associated with each word in the vocabulary.

The parameters are calculated using the formulas:

$$
P(w_i|Ham) = \frac{N_{w_i|Ham} + \alpha}{N_{Ham} + \alpha \cdot N_{Vocabulary}}
$$

$$
P(w_i|Spam) = \frac{N_{w_i|Spam} + \alpha}{N_{Spam} + \alpha \cdot N_{Vocabulary}}
$$

In [25]:
# Initiate parameters
parameters_spam = {unique_word:0 for unique_word in vocabulary}
parameters_ham = {unique_word:0 for unique_word in vocabulary}

# Calculate parameters
for word in vocabulary:
    n_word_given_spam = spam_messages[word].sum()   # spam_messages already defined in a cell above
    p_word_given_spam = (n_word_given_spam + alpha) / (n_spam + alpha*n_vocabulary)
    parameters_spam[word] = p_word_given_spam
    
    n_word_given_ham = ham_messages[word].sum()   # ham_messages already defined in a cell above
    p_word_given_ham = (n_word_given_ham + alpha) / (n_ham + alpha*n_vocabulary)
    parameters_ham[word] = p_word_given_ham


## Classifying a New Message

Now that we calculated all the constants and parameters, we can start creating the spam filter.

In [26]:
import re

def classify(message):
    '''
    message: a string
    '''
    
    message = re.sub('\W', ' ', message)
    message = message.lower().split()
    
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham

    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]
            
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]
            
    print('P(Spam|message):', p_spam_given_message)
    print('P(Ham|message):', p_ham_given_message)
    
    if p_ham_given_message > p_spam_given_message:
        print('Label: Ham')
    elif p_ham_given_message < p_spam_given_message:
        print('Label: Spam')
    else:
        print('Equal proabilities, have a human classify this!')

In [27]:
classify('WINNER!! This is the secret code to unlock the money: C3421.')

P(Spam|message): 1.3481290211300841e-25
P(Ham|message): 1.9368049028589875e-27
Label: Spam


In [28]:
classify("Sounds good, Tom, then see u there")

P(Spam|message): 2.4372375665888117e-25
P(Ham|message): 3.687530435009238e-21
Label: Ham


## Measuring the Spam Filter's Accuracy

After we have created a spam filter, we will now try to determine how well the spam filter does on our test set of 1,114 messages.

The output of the algorithm is a classification label for every message in our test set, which we will then compare with the actual label.

We'll start by writing a function that returns classification labels instead of printing them.

In [29]:
def classify_test_set(message):

    message = re.sub('\W', ' ', message)
    message = message.lower()
    message = message.split()

    p_spam_given_message = p_spam
    p_ham_given_message = p_ham

    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]

        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]

    if p_ham_given_message > p_spam_given_message:
        return 'ham'
    elif p_spam_given_message > p_ham_given_message:
        return 'spam'
    else:
        return 'needs human classification'

After we have a function that returns labels instead of printing them, we can use itto create a new column:

In [30]:
test_set['predicted'] = test_set['SMS'].apply(classify_test_set)
test_set.head()

Label                                                SMS predicted
0   ham          later i guess  i needa do mcat study too        ham
1   ham             but i haf enuff space got like 4 mb          ham
2  spam  had your mobile 10 mths  update to latest oran...      spam
3   ham  all sounds good  fingers   makes it difficult ...       ham
4   ham  all done  all handed in  don t know if mega sh...       ham

At this point we can compare the predicted values with the actual values to measure how good our spam filter is.
We will use accuracy as a metric:

$$
Accuracy = \frac {number\ of\ correctly \ classifies\ messages}{total\ number\ of \ classified\ messages}
$$


In [32]:
#Measuring the accuracy of the spam filter
correct = 0
total = test_set.shape[0]

for row in test_set.iterrows():
    row = row[1]
    if row['Label'] == row['predicted']:
        correct += 1
        
print('Correct:', correct)
print('Incorrect:', total - correct)
print('Accuracy:', correct/total)

Correct: 1100
Incorrect: 14
Accuracy: 0.9874326750448833


The accuracy of the spam filter shows 98%, which is very accurate (taking into consideration the fact that we were agreeing on 80% accuracy before starting the project).

## Next Steps

In this project, we managed to build a spam filter for SMS messages using the multinomial Naive Bayes algorithm. The filter had an accuracy of 98.74% on the test set we used, which is a pretty good result. Our initial goal was an accuracy of over 80%, and we managed to do way better than that.

Next steps include:
    
* Analyze the 14 messages that were classified incorrectly and try to figure out why the algorithm classified them incorrectly
* Make the filtering process more complex by making the algorithm sensitive to letter case